src/main/java/com/icthh/xm/tmf/ms/communication/service/mail/TenantEmailTemplateService.java

```diff
@@ -19,69 +21,112 @@
+    }
+
+    @Override
+    public void onInit(String key, String config) {
+        if (isListeningConfiguration(key)) {
+            onRefresh(key, config);
         }
-        String tenantKeyValue = matcher.group(TENANT_NAME);
-        String langKey = matcher.group(LANG_KEY);
-        String templateName = matcher.group(TEMPLATE_NAME);
+    }
 
-        String templateKey = EmailTemplateUtil.emailTemplateKey(TenantKey.valueOf(tenantKeyValue),
-            templateName, langKey);
+    private String updateTemplates(Map<String, String> emailTemplates, String pathPattern, String key, String config) {
+        Map<String, String> pathVariables = matcher.extractUriTemplateVariables(pathPattern, key);
+        String templatePath = matcher.extractPathWithinPattern(pathPattern, key);
+        String langKey = pathVariables.get(LANG_KEY);
+        String templateFileName = String.format(FILE_PATTERN, pathVariables.get(LANG_KEY));
+        templatePath = templatePath.substring(0, templatePath.lastIndexOf(templateFileName));
+        String tenantKeyValue = pathVariables.get(TENANT_KEY);
+
+        String templateKey = EmailTemplateUtil.emailTemplateKey(TenantKey.valueOf(tenantKeyValue), templatePath, langKey);
 
         if (StringUtils.isBlank(config)) {
             emailTemplates.remove(templateKey);
-            templateLoader.removeTemplate(templateKey);
-            log.info("Email template '{}' with locale {} for tenant '{}' was removed", templateName,
-                            langKey, tenantKeyValue);
+            log.info("Email template '{}' with locale {} for tenant '{}' was removed", templatePath,
+                langKey, tenantKeyValue);
         } else {
             emailTemplates.put(templateKey, config);
-            templateLoader.putTemplate(templateKey, config);
-            log.info("Email template '{}' with locale {} for tenant '{}' was updated", templateName,
-                            langKey, tenantKeyValue);
+            log.info("Email template '{}' with locale {} for tenant '{}' was updated", templatePath,
+                langKey, tenantKeyValue);
         }
+
+        return templateKey;
     }
 
-    @Override
-    public boolean isListeningConfiguration(String updatedKey) {
-        return pattern.matcher(updatedKey).matches();
+    private void updateTemplateLoaderByEmailTemplate(String templateKey, String config) {
+        if (!customEmailTemplates.containsKey(templateKey)) {
+            if (StringUtils.isBlank(config)) {
+                templateLoader.removeTemplate(templateKey);
+            } else {
+                templateLoader.putTemplate(templateKey, config);
+            }
+        }
     }
 
-    @Override
-    public void onInit(String key, String config) {
-        if (isListeningConfiguration(key)) {
-            onRefresh(key, config);
+    private void updateTemplateLoaderByCustomEmailTemplate(String templateKey, String config) {
+        if (StringUtils.isBlank(config)) {
```

Comment:
difficult code better to just use `getTemplateOverrideable().ifPresentOrElse((config) -> templateLoader.putTemplate(templateKey, config), () -> templateLoader.removeTemplate(templateKey))`